pandas/core/reshape/melt.py

```diff
@@ -115,7 +116,17 @@ def melt(
         # asanyarray will keep the columns as an Index
         mdata[col] = np.asanyarray(frame.columns._get_level_values(i)).repeat(N)
 
-    return frame._constructor(mdata, columns=mcolumns)
+    result = frame._constructor(mdata, columns=mcolumns)
+
+    if index:
+        new_index = np.tile(frame.index, K)
+
+        if isinstance(frame.index, MultiIndex):  # Index of tuples otherwise
```

Comment:
you might be able to use .from_arrays instead

Reply:
It seems the natural "arrangement" for MultiIndex values is in tuples form. There is the `codes`-property but that doesn't work because it's not the original index values. I think using from_tuples is preferable in this case.